### DISCUSSION TITLE
Is it possible to force HTTP/2 for outbound connections over TLS?

### DISCUSSION THREAD

**Discussion by pozernishku** [mitmproxy/mitmproxy#5559] [Q&A] [answered]
My question as simple as this option in `Burp`:

<img width="703" alt="Снимок экрана 2022-08-27 в 08 23 01" src="https://user-images.githubusercontent.com/14317371/187016126-3d2df893-31c0-478b-8f9a-af47326c8e23.png">

I mean, is it possible to set up the `mitmproxy` the same way as `Burp`. For example I send `HTTP/1.1` request using `curl` to `example.com`. It goes through the `mitmproxy` and turns out the outbound connection to the server to `HTTP/2`?

<img width="1021" alt="Снимок экрана 2022-08-27 в 08 49 44" src="https://user-images.githubusercontent.com/14317371/187016937-3628ce0a-2a10-4a5b-9b88-478ebc70289c.png">

**Comment by mhils** [comment, marked as answer]
We try to keep the HTTP version between both connections (client-proxy and proxy-server) in sync so that adding mitmproxy does not magically add or remove bugs. Technically mitmproxy is perfectly capable of translating between versions. If needed, you can force a specific HTTP version with a custom addon that modifies the TLS ALPN offers:

```
def server_connect(data):
    data.server.alpn_offers = [b"h2"]
```

**Reply by pozernishku** [reply]
Thank you for the clear answer, I learned a lot 💪 
I've tried the simplest addons from [here](https://docs.mitmproxy.org/stable/addons-overview/) and they magically work, great! 
I am still thinking how to solve my issue. I've tried your code and found that the HTTP version is still 1.1 😢  (the same as my request).
It looks like I missed something or the code you've presented is just not enough to change the version (?)
I will be really pleased to see some more tips. Thank you!

<img width="584" alt="Снимок экрана 2022-08-28 в 23 02 39" src="https://user-images.githubusercontent.com/14317371/187092404-f2faa0e4-4484-4988-9c0f-3a509f88e793.png">

**Reply by mhils** [reply]
"Client Connection" refers to the connection between client and proxy. Unless I misunderstand your initial request, you want HTTP/2 on the server connection?

**Reply by pozernishku** [reply]
Yes, here:

<img width="676" alt="Снимок экрана 2022-08-29 в 23 37 38" src="https://user-images.githubusercontent.com/14317371/187294942-ec518f37-9d2d-4783-a307-d503c90d8259.png">

**Reply by mhils** [reply]
Yes. The screenshot you posted earlier only shows that there is HTTP/1.1 between client and proxy. It doesn't state anything for the proxy - server connection.

**Reply by pozernishku** [reply]
Great, thank you!

How can I check the proxy - server connection, then?

**Reply by pozernishku** [reply]
I got it! You cannot see `Server Connection` in mitmproxy upstream mode.
